broadcast/src/lib.rs

```diff
@@ -0,0 +1,112 @@
+//! Replication of messages across a network.
+
+use bytes::Bytes;
+use commonware_cryptography::Array;
+use futures::channel::oneshot;
+use std::future::Future;
+
+pub mod linked;
+
+/// Proof is a blob that attests to some data.
+pub type Proof = Bytes;
+
+/// Broadcaster is the interface responsible for replication of messages across a network.
+pub trait Broadcaster: Clone + Send + 'static {
+    /// Digest is an arbitrary hash digest.
+    type Digest: Array;
+
+    /// Attempt to broadcast a digest to the network.
+    ///
+    /// Returns a future that resolves to a boolean indicating success.
+    fn broadcast(
+        &mut self,
+        payload: Self::Digest,
+    ) -> impl Future<Output = oneshot::Receiver<bool>> + Send;
+}
+
+/// Application is the interface responsible for processing messages received from the network.
+pub trait Application: Clone + Send + 'static {
+    /// Context is metadata provided by the broadcast engine to associated with a given payload.
+    /// This could include things like the public key of the sequencer.
+    type Context;
+
+    /// Digest is an arbitrary hash digest.
+    type Digest: Array;
+
+    /// Verify a proposed payload received from the network.
+    ///
+    /// Returns a future that resolves to a boolean indicating success.
+    /// Part of verification is ensuring that the data is made available, for example by storing it in a database.
+    fn verify(
+        &mut self,
+        context: Self::Context,
+        payload: Self::Digest,
+    ) -> impl Future<Output = oneshot::Receiver<bool>> + Send;
+}
+
+/// Collector is the interface responsible for handling notifications of broadcasted payloads.
+pub trait Collector: Clone + Send + 'static {
+    /// Context is metadata provided by the broadcast engine to associated with a given payload.
+    /// This could include things like the public key of the sequencer.
+    type Context;
+
+    /// Digest is an arbitrary hash digest.
+    type Digest: Array;
+
+    /// Emit that a payload has been successfully broadcasted.
+    /// This is used to acknowledge that the payload has been "received" by the network,
+    /// for example that it has been successfully gossiped to a threshold of validators.
+    fn acknowledged(
```

Comment:
In consensus, `proof` goes before other arguments.

Do we need to pass `context` here? Seems like you could probably get what you need by parsing `proof`? (Side note, I'm working through this "reporter" interface rn with najeal and not in love with what we have anyways: https://github.com/commonwarexyz/monorepo/pull/462#discussion_r1945758182

Reply:
Will just remove `context` for now and call it a day. Will wait for a larger overhaul of the proof stuff to land to change this.

A potential change cloud look like the following:

old:
```
type Proof = Bytes;
pub trait Collector {
    type Context;
    type Digest: Array;
    fn acknowledged(
        &mut self,
        context: Self::Context,
        payload: Self::Digest,
        proof: Proof,
    ) -> impl Future<Output = ()> + Send;
}
```

new:
```
pub trait Collector {
    type Proof;
    type Digest: Array;
    fn acknowledged(
        &mut self,
        proof: Self::Proof,
        payload: Self::Digest, // maybe delete this too?
    ) -> impl Future<Output = ()> + Send;
}
```